src/main/java/org/wso2/carbon/inbound/kafka/deserializer/ErrorHandlingDeserializer.java

```diff
@@ -0,0 +1,189 @@
+            setupDelegate(configs, isKey ? KafkaConstants.KEY_DELEGATE_DESERIALIZER
+                    : KafkaConstants.VALUE_DELEGATE_DESERIALIZER);
+        }
+        if (this.delegate == null) {
+            LOGGER.error("No delegate deserializer configured");
+        }
+        this.delegate.configure(configs, isKey);
+        this.isForKey = isKey;
+    }
+
+    @Override
+    public T deserialize(String topic, byte[] data) {
+
+        try {
+            return this.delegate.deserialize(topic, data);
+        } catch (Exception e) {
+            Iterable<Header> recordHeaders = null;
+            Headers headers = new RecordHeaders(recordHeaders);
+            deserializationException(topic, headers, data, e, this.isForKey);
+            return null;
+        }
+    }
+
+    @Override
+    public T deserialize(String topic, Headers headers, byte[] data) {
+
+        try {
+            if (this.isForKey) {
+                headers.remove(KafkaConstants.KEY_DESERIALIZER_EXCEPTION_HEADER);
+            }
+            else {
+                headers.remove(KafkaConstants.VALUE_DESERIALIZER_EXCEPTION_HEADER);
+            }
+            return this.delegate.deserialize(topic, headers, data);
+        } catch (Exception e) {
+            deserializationException(topic, headers, data, e, this.isForKey);
+            return null;
+        }
+    }
+
+    @Override
+    public void close() {
+
+        Deserializer.super.close();
+    }
+
+    private void setupDelegate(Map<String, ?> configs, String configKey) {
+        if (configs.containsKey(configKey)) {
+            try {
+                Object value = configs.get(configKey);
+                ClassLoader classLoader = getDefaultClassLoader();
+                if (classLoader != null) {
+                    Class<?> clazz = classLoader.loadClass(value.toString());
+                    this.delegate = setupDelegate(clazz.getDeclaredConstructor().newInstance());
+                } else {
+                    throw new SynapseException("Could not setup the delegate as the ClassLoader was not accessible.");
+                }
+            } catch (Exception e) {
+                throw new SynapseException("Could not setup the delegate.", e);
+            }
+        }
+    }
+
+    private Deserializer<T> setupDelegate(Object delegate) {
+        if (!(delegate instanceof Deserializer)) {
+            throw new SynapseException("'delegate' must be a 'Deserializer'");
+        }
+        return (Deserializer<T>) delegate;
+    }
+
+    /**
+     * Populate the record headers with a serialized {@link DeserializationException}.
+     * @param topic the topic.
+     * @param headers the headers.
+     * @param data the data.
+     * @param ex the exception.
+     * @param isForKeyArg true if this is a key deserialization problem, otherwise value.
+     */
+    private static void deserializationException(String topic, Headers headers, byte[] data, Exception ex,
```

Comment:
Shall we rename this method to something like `populateRecordHeadersWithException`? `deserializationException` looks ambiguous.

Reply:
https://github.com/wso2-extensions/esb-inbound-kafka/pull/59/commits/ff12250e6c49f14d72ef658730c55f49849ce8a4